### PR TITLE
Add Sensible Items for Gen 3+

### PR DESCRIPTION
**Pull request by spaceonaut**
This PR implements the sensible items logic for Gens 3+.

I chose not to add more methods to RomHandler and AbstractRomHandler (e.g. functions to get the various breakdowns of items) because that seemed like 1) a leaky abstraction and 2) would make the code pretty hard to follow.

I chose to make some items (e.g. species specific items, weakness-berries, type-boosting items if a pokemon knows more than 1 damaging move of that type) more common rather than everything the same rarity.

I also made a few choices around flameOrb and toxicOrb (e.g. give them to Pokemon with guts or Fling/Trick etc).

Let me know what y'all think of those choices as well as the implementations.

### CODE REVIEW

File: src/com/dabomstew/pkrandom/constants/Gen3Constants.java [hunk 1]
```diff
@@ -236,7 +239,62 @@
...
+            cheriBerry, chestoBerry, pechaBerry, rawstBerry, aspearBerry, leppaBerry,
+            oranBerry, persimBerry, lumBerry, sitrusBerry, ganlonBerry, salacBerry,
+            // An NPC pokemon's nature is generated randomly with IVs during gameplay. Therefore, we include
+            // all these berries, even though they might confuse the opponent's pokemon if the nature is wrong.
+            figyBerry, wikiBerry, magoBerry, aguavBerry, iapapaBerry,
+            apicotBerry, lansatBerry, starfBerry, berryJuice, whiteHerb, mentalHerb
+    ));
+
+    public static final List<Integer> generalPurposeItems = Collections.unmodifiableList(Arrays.asList(
+            brightPowder, quickClaw, choiceBand, kingsRock, focusBand,
```
Comment (tom-overton) [inline src/com/dabomstew/pkrandom/constants/Gen3Constants.java:255]: Choice Band shouldn't be here, since you add it separately in Gen3RomHandler if a move is physical

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java [hunk 2]
```diff
@@ -150,18 +154,166 @@
...
+            // An NPC pokemon's nature is generated randomly with IVs during gameplay. Therefore, we include
+            // all these berries, even though they might confuse the opponent's pokemon if the nature is wrong.
+            figyBerry, wikiBerry, magoBerry, aguavBerry, iapapaBerry,
+            chilanBerry, apicotBerry, lansatBerry, starfBerry, enigmaBerry, micleBerry, custapBerry,
+            jabocaBerry, rowapBerry, berryJuice, whiteHerb, mentalHerb, focusSash));
+
+    public static final List<Integer> generalPurposeItems = Collections.unmodifiableList(Arrays.asList(
+            brightPowder, quickClaw, choiceBand, kingsRock, focusBand,
+            scopeLens, leftovers, shellBell, laxIncense, wideLens, expertBelt, lifeOrb, zoomLens,
+            laggingTail, destinyKnot, stickyBarb, shedShell, fullIncense, razorClaw, razorFang));
```
Comment (tom-overton) [inline src/com/dabomstew/pkrandom/constants/Gen4Constants.java:191]: Choice Band shouldn't be here for the same reason as Gen 3. Additionally, I'm not sure if I would call Lagging Tail "general purpose"

File: src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java [hunk 3]
```diff
@@ -3461,4 +3462,75 @@ public BufferedImage getMascotImage() {
...
+            if (tp.pokemon.primaryType == Type.POISON || tp.pokemon.secondaryType == Type.POISON) {
+                items.add(Gen4Constants.blackSludge);
+            }
+            List<Integer> speciesItems = Gen5Constants.speciesBoostingItems.get(tp.pokemon.number);
+            if (speciesItems != null) {
+                for (int i = 0 ; i < 8; i ++) {  // Increase the likelihood of using species specific items.
+                    items.addAll(speciesItems);
+                }
+            }
+            if (!tp.pokemon.evolutionsFrom.isEmpty()) {
```
Comment (tom-overton) [inline src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java:3530]: Both you and Ajarmar can disagree with me here, but I feel like earlygame Pokemon having Eviolite is just super annoying. Maybe gating this behind a level (25? 30? 35?) would make more sense. Again, just sort of spitballing here, feel free to say "no"

**Review by tom-overton — CHANGES_REQUESTED** [review]

File: src/com/dabomstew/pkrandom/constants/Gen3Constants.java [hunk 4]
```diff
@@ -236,7 +239,62 @@
...
-        return list;
+        return Collections.unmodifiableList(list);
+    }
+
+    public static final List<Integer> generalPurposeConsumableItems = Collections.unmodifiableList(Arrays.asList(
+            cheriBerry, chestoBerry, pechaBerry, rawstBerry, aspearBerry, leppaBerry,
+            oranBerry, persimBerry, lumBerry, sitrusBerry, ganlonBerry, salacBerry,
+            // An NPC pokemon's nature is generated randomly with IVs during gameplay. Therefore, we include
+            // all these berries, even though they might confuse the opponent's pokemon if the nature is wrong.
+            figyBerry, wikiBerry, magoBerry, aguavBerry, iapapaBerry,
```
[the remark this answers is not in the document]
Comment (Ajarmar) [reply src/com/dabomstew/pkrandom/constants/Gen3Constants.java:250]: I agree, before Gen 7 it would make more sense to not include these for "Sensible Items".

File: src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java [hunk 5]
```diff
@@ -3461,4 +3462,75 @@ public BufferedImage getMascotImage() {
...
+            Integer berry = Gen5Constants.weaknessReducingBerries.get(entry.getKey());
+            if (entry.getValue() == Effectiveness.DOUBLE) {
+                items.add(berry);
+            } else if (entry.getValue() == Effectiveness.QUADRUPLE) {
+                for (int i = 0; i < 8; i ++) {  // Increase the likelihood of protecting against 4x weak.
+                    items.add(berry);
+                }
+            }
+        }
+        if (byType.get(Type.GROUND) != Effectiveness.ZERO && byType.get(Type.GROUND) != Effectiveness.QUARTER) {
```
[the remark this answers is not in the document]
Comment (Ajarmar) [reply src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java:3512]: Yeah, the Air Balloon is similar to the weakness-reducing berries so I think it should follow the same rules as those.

File: src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java (hunk 3 above)
Comment (Ajarmar) [reply src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java:3530]: I'd have to agree - Eviolite stands out as a _very_ strong item, especially very early in the game. That said, since it's rare for non-evolved Pokemon to appear late in the game, it shouldn't have too high of a level gate. I think 25 seems reasonable.

File: src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java [hunk 6]
```diff
@@ -3413,4 +3413,46 @@ public BufferedImage getMascotImage() {
...
+                items.add(Gen3Constants.petayaBerry);
+                if (!consumableOnly) {
+                    items.addAll(Gen3Constants.typeBoostingItems.get(move.type));
+                }
+            }
+        }
+        if (!consumableOnly) {
+            List<Integer> speciesItems = Gen3Constants.speciesBoostingItems.get(tp.pokemon.number);
+            if (speciesItems != null) {
+                for (int i = 0 ; i < 8; i ++) {  // Increase the likelihood of using species specific items.
```
Comment (Ajarmar) [inline src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java:3451]: I think it's reasonable to boost the probability for certain items to appear, but is 8 the appropriate multiplier for this? Not just for the species specific ones, but also the weakness reducing berries in Gen 4/5. I would put it maybe at 4, which would still make them significantly more likely but not overwhelmingly so. Feel free to give your own thoughts, though.

File: src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java [hunk 7]
```diff
@@ -4294,4 +4294,67 @@ public BufferedImage getMascotImage() {
...
+            }
+        }
+
+        if (!consumableOnly) {
+            if (tp.ability == Gen4Constants.guts) {
+                items.add(Gen4Constants.flameOrb);
+                items.add(Gen4Constants.toxicOrb);
+            }
+            if (tp.pokemon.primaryType == Type.POISON || tp.pokemon.secondaryType == Type.POISON) {
+                items.add(Gen4Constants.blackSludge);
```
Comment (Ajarmar) [inline src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java:4349]: Wouldn't it make more sense for Black Sludge to be put in the "Type Boosting Items" map, instead of giving it its own case like this?

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java [hunk 8]
```diff
@@ -150,18 +154,166 @@
...
+        map.put(Type.GROUND, shucaBerry);
+        map.put(Type.FLYING, cobaBerry);
+        map.put(Type.PSYCHIC, payapaBerry);
+        map.put(Type.BUG, tangaBerry);
+        map.put(Type.ROCK, chartiBerry);
+        map.put(Type.GHOST, kasibBerry);
+        map.put(Type.DRAGON, habanBerry);
+        map.put(Type.DARK, colburBerry);
+        map.put(Type.STEEL, babiriBerry);
+        // Note: Nothing is weak to normal, so the chilanBerry is "general purpose"
```
Comment (Ajarmar) [inline src/com/dabomstew/pkrandom/constants/Gen4Constants.java:294]: Should the Chilan Berry be "General Purpose"? It's useless on Ghost types and mediocre on types that resist Normal. I think it warrants having a special case, where it's only put on Pokemon that take neutral damage from Normal.

**Review by Ajarmar — CHANGES_REQUESTED** [review]

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java (hunk 8 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen4Constants.java]: Done.

File: src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java (hunk 7 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java:4352]: typeBoosting item cares about the move type. Black sludge cares about the pokemon type.

File: src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java (hunk 6 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java:3447]: The item pool is pretty large. Even with the boost of 8x, the supereffective berries only show up about a dozen times each across the 1304 pokemon in Pokemon Black. In soulsilver, the numbers creep up into the high teens or so (out of 1776 pokemon). The chilan berry (unboosted and only for neutral effective pokemon) shows up 20-30 times (as a comparison). Bright powder (an arbitrary "general purpose" item) also shows up 20-30 times.

For species specific items, with the 8x boost, they typically show up 0 or 1 times when giving random items to all trainer pokemon in the game.

In short, with the 8x multiplier, the weakness reducing berries (8x only applies to double weakness) show up less often than a general purpose item. With the 8x multiplier, species specific items are now likely to happen, but only once or twice across an entire run. I played with a few multipliers until this one "felt" right.

File: src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java [hunk 9]
```diff
@@ -3413,4 +3413,46 @@ public BufferedImage getMascotImage() {
     public List<Integer> getAllConsumableHeldItems() {
         return Gen3Constants.consumableHeldItems;
     }
+
+    private static final Set<Type> physicalTypes = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+            Type.NORMAL, Type.FIGHTING, Type.POISON, Type.GROUND, Type.FLYING, Type.BUG,
+            Type.ROCK, Type.GHOST, Type.STEEL)));
```
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java]: Done.

File: src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java (hunk 3 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java]: I added a level cap of 20 (same as Shedninja, right?) I ran the randomizer for Black giving sensible items to every trainer pokemon. Out of the 1304 trainer pokemon, 6 of them were given Eviolite. 

I then ran it again with "Ironmon" difficulty settings (+50% foe level, fully evolved at lvl 30) and eviolite showed up 2 times.

I tried it again with the limit at level 25. Without modifying trainer level or evolutions, eviolite showed up 6 times again. With "Ironmon" difficulty settings, it showed up 0 times.

To me, level 20 feels fine, given how infrequently it shows up on average. What do y'all think with the above frequencies I outlined?

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java (hunk 2 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen4Constants.java]: Done. Removed laggingTail and fullIncense from general purpose and moved them to moveBoostingItems.

File: src/com/dabomstew/pkrandom/constants/Gen3Constants.java (hunk 1 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen3Constants.java]: Done.

File: src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java]: Done.

File: src/com/dabomstew/pkrandom/romhandlers/AbstractRomHandler.java [hunk 10]
```diff
@@ -1930,6 +1930,9 @@ public void randomizeTrainerHeldItems(Settings settings) {
         boolean giveToImportantPokemon = settings.isRandomizeHeldItemsForBossTrainerPokemon();
         boolean giveToRegularPokemon = settings.isRandomizeHeldItemsForBossTrainerPokemon();
         boolean highestLevelOnly = settings.isHighestLevelGetsItemsForTrainers();
+
+        List<Move> moves = this.getMoves();
+        Map<Integer, List<MoveLearnt>> movesets = this.getMovesLearnt();
```
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/AbstractRomHandler.java:1935]: I have it here because I observed noticeable added latency when running locally and attributed it to fetching the moves and movesets repeatedly. Calling these two functions takes a few milliseconds and since it's static data, it didn't make sense to me to call it inside a double for loop. Rough numbers: 3 ms * 600 trainers * 2 pokes / trainer = 3 extra seconds of latency.

To me, that seems like an easy win to keep the moves/movesets here. Notice that we don't throw them away until we've randomized all the trainers and this function only gets called when at least some trainers get randomized.

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java [hunk 11]
```diff
@@ -150,18 +154,166 @@
...
+        map.put(aquaRing, Arrays.asList(bigRoot));
+        map.put(drainPunch, Arrays.asList(bigRoot));
+        map.put(dreamEater, Arrays.asList(bigRoot));
+        map.put(gigaDrain, Arrays.asList(bigRoot));
+        map.put(ingrain, Arrays.asList(bigRoot));
+        map.put(leechLife, Arrays.asList(bigRoot));
+        map.put(leechSeed, Arrays.asList(bigRoot));
+        map.put(megaDrain, Arrays.asList(bigRoot));
+
+        return Collections.unmodifiableMap(map);
```
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen4Constants.java:272]: Done.

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java [hunk 12]
```diff
@@ -150,18 +154,166 @@
...
+        map.put(fly, Arrays.asList(powerHerb));
+        map.put(razorWind, Arrays.asList(powerHerb));
+        map.put(skullBash, Arrays.asList(powerHerb));
+        map.put(skyAttack, Arrays.asList(powerHerb));
+        map.put(solarBeam, Arrays.asList(powerHerb));
+
+        map.put(fling, Arrays.asList(toxicOrb, flameOrb, ironBall));
+
+        map.put(trick, Arrays.asList(toxicOrb, flameOrb));
+        map.put(switcheroo, Arrays.asList(toxicOrb, flameOrb));
```
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen4Constants.java]: Done.

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java [hunk 13]
```diff
@@ -150,18 +154,166 @@
...
+    // Abilties with positive item synergy
+    // https://bulbapedia.bulbagarden.net/wiki/Ability#List_of_Abilities
+    public static final int guts = 62;
+
+    public static final List<Integer> generalPurposeConsumableItems = Collections.unmodifiableList(Arrays.asList(
+            cheriBerry, chestoBerry, pechaBerry, rawstBerry, aspearBerry, leppaBerry,
+            oranBerry, persimBerry, lumBerry, sitrusBerry, ganlonBerry, salacBerry,
+            // An NPC pokemon's nature is generated randomly with IVs during gameplay. Therefore, we include
+            // all these berries, even though they might confuse the opponent's pokemon if the nature is wrong.
+            figyBerry, wikiBerry, magoBerry, aguavBerry, iapapaBerry,
```
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen4Constants.java]: Done.

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java [hunk 14]
```diff
@@ -124,15 +124,19 @@
...
             heatRock = 0x11c, dampRock = 0x11d, gripClaw = 0x11e, choiceScarf = 0x11f, stickyBarb = 0x120,
             shedShell = 0x127, bigRoot = 0x128, choiceSpecs = 0x129, flamePlate = 0x12a, splashPlate = 0x12b,
             zapPlate = 0x12c, meadowPlate = 0x12d, iciclePlate = 0x12e, fistPlate = 0x12f, toxicPlate = 0x130,
             earthPlate = 0x131, skyPlate = 0x132, mindPlate = 0x133, insectPlate = 0x134, stonePlate = 0x135,
             spookyPlate = 0x136, dracoPlate = 0x137, dreadPlate = 0x138, ironPlate = 0x139, oddIncense = 0x13a,
-            rockIncense = 0x13b, fullIncense = 0x13c, waveIncense = 0x13d, roseIncense = 0x13e, pureIncense = 0x140,
+            rockIncense = 0x13b, fullIncense = 0x13c, waveIncense = 0x13d, roseIncense = 0x13e,
             razorClaw = 0x146, razorFang = 0x147;
+    // species-specific held items.
+    public static final int adamantOrb = 0x087, lustrousOrb = 0x088, soulDew = 0x0e1, deepSeaTooth = 0x0e2,
```
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen4Constants.java:138]: Yes - I didn't want to break anything and having it for some games but not others seemed difficult.

File: src/com/dabomstew/pkrandom/constants/Gen3Constants.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen3Constants.java]: Done. I've been trying out Platinum with the code here and while it's nice to occasionally be bailed out by a confusion berry, it's probably better to leave it out until Gen 7 where healing 1/2 is worth the risk.

File: src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java (hunk 7 above)
Comment (Ajarmar) [reply src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java:4352]: Right, my bad.

File: src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java (hunk 6 above)
Comment (Ajarmar) [reply src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java:3447]: Rather than looking at how many times the boosted-probability items appear in total, I think the more interesting metric to look at is: for Pokemon that are affected by this, how high is the probability that they will get the boosted-probability item? A player probably won't keep track of how many total Yache Berries they've run into in a single run, but if they encounter two Dragonites in the same run that both have Yache Berries they will probably notice. It shouldn't be too predictable.

I tried running a few tests on Emerald, Soul Silver and Black, to see how large the item pools/probabilities were for Pokemon where boosted-probability items were picked:

Emerald: ~35-50 (~16-23%)
Soul Silver: ~60-80 (~10-13%)
Soul Silver (consumable only): ~35-50 (~16-23%)
Black: ~65-95 (~8-12%)
Black (consumable only): ~40-55 (~15-20%)

This is not accounting for the Pokemon that have two quad weaknesses, where the percentages would be doubled (although their pools would be on the larger end so we wouldn't be looking at 46% for SS/consumable only - but 32% is still pretty high). [Here are the raw numbers from a single run of each, if you're interested.](https://pastebin.com/jLXhk7hr)

I think some of these are fine while others may warrant adjusting. 20% is too high in my opinion. Tweaking these on a per-generation basis as well as depending on the "Consumable Only" setting seems reasonable to me. I would suggest something like:

Gen 3: 6
Gen 4 (consumable only): 6
Gen 5 (consumable only): 6

What do you think?

File: src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java (hunk 3 above)
Comment (tom-overton) [reply src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java]: Yes, that's the same level used for "No Early Wonder Guard". I'd be interested in seeing what the probability is *for unevolved Pokemon*. So in your example, 6/1304 trainer Pokemon got it, but how many of those 1304 Pokemon were eligible to get it in the first place? But overall, this probability seems fine to me.

**Review by tom-overton — COMMENTED** [review]
Looks much better now, thanks. I think the remaining things to do are:
- Iron out the probability stuff
- Gen 6/7 implementation

File: src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java (hunk 6 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen3RomHandler.java:3447]: That is excellent analysis. I will implement your suggestions. I agree that 20ish percent is a bit high for any given item to appear, and 10ish percent is a bit of a nicer middle ground - common, but not too common.

File: src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java (hunk 3 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java]: Ran it again for Pokemon Black with level cap 20
 - normal levels, no early evolution: 514 out of 1304 were eligible for eviolite; 4 were given it (~1%).
 - "ironmon" difficulty: 51 out of 1304 were eligible for eviolite; 1 was given it (~2%)

**Comment by spaceonaut** [comment]
I'll see about getting to Gen 6/7 by the end of this weekend.

File: src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java (hunk 3 above)
Comment (tom-overton) [reply src/com/dabomstew/pkrandom/romhandlers/Gen5RomHandler.java]: Yeah those seem like acceptable odds to me. At the very least, it's clear that level 30 is way too high, so if these odds are too high, then we can just bump it to 25.

**Comment by spaceonaut** [comment]
Gens 6 and 7 added. I rebased to make use of the new Abilities.java as well.

File: src/com/dabomstew/pkrandom/romhandlers/Gen7RomHandler.java [hunk 15]
```diff
@@ -3196,4 +3198,92 @@ public AreaData() {
...
+                    items.addAll(Gen7Constants.typeBoostingItems.get(move.type));
+                    items.add(Gen4Constants.wiseGlasses);
+                    items.add(Gen4Constants.choiceSpecs);
+                }
+            }
+            if (!consumableOnly && Gen7Constants.moveBoostingItems.containsKey(moveIdx)) {
+                items.addAll(Gen7Constants.moveBoostingItems.get(moveIdx));
+            }
+        }
+        if (hasAtLeastOneDamagingMove) {
```
Comment (tom-overton) [inline src/com/dabomstew/pkrandom/romhandlers/Gen7RomHandler.java:3242]: Same thing as what I said in Gen6RomHandler about Assault Vest

File: src/com/dabomstew/pkrandom/constants/Gen7Constants.java [hunk 16]
```diff
@@ -210,13 +216,67 @@ public static int getPokemonCount(int romType) {
...
-        list.addAll(Gen6Constants.allHeldItems);
+        List<Integer> list = new ArrayList<>(Gen6Constants.allHeldItems);
         list.addAll(Arrays.asList(adrenalineOrb, electricSeed, psychicSeed, mistySeed, grassySeed));
         list.addAll(Arrays.asList(terrainExtender, protectivePads));
         return list;
     }
 
+    public static final List<Integer> generalPurposeConsumableItems = initializeGeneralPurposeConsumableItems();
+
+    private static List<Integer> initializeGeneralPurposeConsumableItems() {
```
Comment (tom-overton) [inline src/com/dabomstew/pkrandom/constants/Gen7Constants.java:227]: I was a little surprised not to see the pinch berries here. Does the confusion chance still make it not worth it?

Additionally, while I haven't looked into this very much, it looks like Natures are defined on the trainer Pokemon data structure in Gen 7:
image

We actually read and write the nature in Gen7RomHandler (doesn't look like we *ever* change it though), so *if* I'm understanding this correctly, we could give Gen 7 trainer Pokemon pinch berries that *wouldn't* confuse them. This probably involves a decent amount of effort, though, so I totally understand if you don't want to do it or want to make it a follow-up PR.

**Review by tom-overton — COMMENTED** [review]
Looks pretty good, in fact the worst thing here is a problem of my own creation. When I originally made my "Ban Useless Abilities" commit, I forgot to include a copyright header on Abilities.java, so I force-pushed to fix that. Unfortunately, you rebased before that force-push happened, which is making github spit out a pretty gnarly-looking diff. Can you rebase onto the latest master again to fix that? Sorry about that!

File: src/com/dabomstew/pkrandom/romhandlers/Gen6RomHandler.java [hunk 17]
```diff
@@ -3501,4 +3503,89 @@ public BufferedImage getMascotImage() {
...
+                    items.addAll(Gen6Constants.typeBoostingItems.get(move.type));
+                    items.add(Gen4Constants.wiseGlasses);
+                    items.add(Gen4Constants.choiceSpecs);
+                }
+            }
+            if (!consumableOnly && Gen6Constants.moveBoostingItems.containsKey(moveIdx)) {
+                items.addAll(Gen6Constants.moveBoostingItems.get(moveIdx));
+            }
+        }
+        if (hasAtLeastOneDamagingMove) {
```
[the remark this answers is not in the document]
Comment (Ajarmar) [reply src/com/dabomstew/pkrandom/romhandlers/Gen6RomHandler.java:3547]: Agreed. Could even increase the chance of Assault Vest appearing depending on how many attacking moves the Pokemon has, something like 2/3 attacking moves = 1x, 4 attacking moves = 2x.

**Review by Ajarmar — CHANGES_REQUESTED** [review]

File: src/com/dabomstew/pkrandom/constants/Gen4Constants.java [hunk 18]
```diff
@@ -150,31 +155,185 @@
...
+            cheriBerry, chestoBerry, pechaBerry, rawstBerry, aspearBerry, leppaBerry,
+            oranBerry, persimBerry, lumBerry, sitrusBerry, ganlonBerry, salacBerry,
+            // An NPC pokemon's nature is generated randomly with IVs during gameplay. Therefore, we do not include
+            // the flavor berries because, prior to Gen 7, they aren't worth the risk.
+            apicotBerry, lansatBerry, starfBerry, enigmaBerry, micleBerry, custapBerry,
+            jabocaBerry, rowapBerry, berryJuice, whiteHerb, mentalHerb, focusSash));
+
+    public static final List<Integer> generalPurposeItems = Collections.unmodifiableList(Arrays.asList(
+            brightPowder, quickClaw, kingsRock, focusBand, scopeLens, leftovers, shellBell, laxIncense,
+            wideLens, expertBelt, lifeOrb, zoomLens, destinyKnot, stickyBarb, shedShell, razorClaw, razorFang));
```
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen4Constants.java]: Done.

File: src/com/dabomstew/pkrandom/constants/Gen7Constants.java (hunk 16 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/constants/Gen7Constants.java:227]: I had forgotten about the healing berries. I added them back in. I'm going to hold off on matching the berries to the natures for now.

File: src/com/dabomstew/pkrandom/pokemon/Effectiveness.java [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+         /*POISON */ {NEUTRAL, NEUTRAL, NEUTRAL,  DOUBLE, NEUTRAL, NEUTRAL,    HALF,    HALF, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL,    HALF,    HALF, NEUTRAL,    ZERO, NEUTRAL},
+         /*ICE    */ {NEUTRAL, NEUTRAL,  DOUBLE,  DOUBLE,    HALF,    HALF, NEUTRAL,  DOUBLE, NEUTRAL, NEUTRAL,  DOUBLE, NEUTRAL, NEUTRAL, NEUTRAL,    HALF,    HALF, NEUTRAL},
+         /*STEEL  */ {NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL,    HALF,    HALF,  DOUBLE, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL,    HALF, NEUTRAL, NEUTRAL,  DOUBLE,    HALF, NEUTRAL},
+         /*DARK   */ {NEUTRAL,    HALF, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL,  DOUBLE, NEUTRAL, NEUTRAL, NEUTRAL,  DOUBLE, NEUTRAL, NEUTRAL,    HALF,    HALF},
+    };
+    private static final Effectiveness[][] gen6PlusTable = {
+                       // NORMAL,FIGHTING, FLYING,   GRASS ,   WATER,   FIRE ,   ROCK , GROUND,  PSYCHIC,   BUG  ,  DRAGON,ELECTRIC,   GHOST , POISON,   ICE  ,  STEEL ,  DARK  , FAIRY
+            /*NORMAL */ {NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL,    HALF, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL, NEUTRAL,    ZERO, NEUTRAL, NEUTRAL,    HALF, NEUTRAL, NEUTRAL},
```
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/pokemon/Effectiveness.java]: Done.

File: src/com/dabomstew/pkrandom/romhandlers/Gen6RomHandler.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen6RomHandler.java]: Done. Just going with one Assault Vest chance for 2+.

File: src/com/dabomstew/pkrandom/romhandlers/Gen7RomHandler.java (hunk 15 above)
Comment (spaceonaut) [reply src/com/dabomstew/pkrandom/romhandlers/Gen7RomHandler.java]: Done.

**Review by spaceonaut — COMMENTED** [review]
Feedback addressed.

**Review by tom-overton — APPROVED** [review]
Looks good, thanks for all the work you've put in responding to everything.

I tested this myself and I noticed there's actually something off about how we handle trainer Pokemon abilities, but it needs investigation and changes from our side of things, so we'll fix it independent of this PR. Just don't be surprised if you see something like a Levitate Pokemon holding a Shuca Berry for now; we'll take care of it.

**Review by Ajarmar — APPROVED** [review]
Yeah, this looks good now. I'll merge it in, thanks!
